Assets/Scripts/Wallet/Wallet.cs

```diff
@@ -0,0 +1,65 @@
+﻿namespace Hermitage
+{
+    namespace Economy
+    {
+        using System;
+        using System.Collections.Generic;
+        using System.Linq;
+        using UnityEngine;
+
+        public abstract class Wallet : IWallet
+        {
+            private List<ICoin> coins;
+
+            protected abstract int Capacity { get; }
+
+            public event EventHandler OnCoinAdded;
+            public event EventHandler OnCoinRemoved;
+
+            public bool IsFull => RemainingCapacity <= 0;
+            public int Value => coins.Sum(c => c.Value);
+            public int RemainingCapacity => Capacity - coins.Count;
+
+            protected Wallet()
+            {
+                coins = new List<ICoin>();
+            }
+
+            public bool CanAfford(int cost) => cost <= Value;
+
+            public ICoin[] Charge(int cost)
+            {
+                List<ICoin> payments = new List<ICoin>();
+
+                foreach (ICoin coin in coins.OrderByDescending(c => c.Value).ToList())
+                {
+                    if (coin.Value > cost - payments.Sum(c => c.Value))
+                    {
+                        continue;
+                    }
+
+                    coins.Remove(coin);
+                    payments.Add(coin);
+                }
+
+                int remainingCost = cost - payments.Sum(c => c.Value);
+                ICoin payment = coins.OrderBy(c => c.Value).FirstOrDefault(c => c.Value >= remainingCost);
+                coins.Remove(payment);
+                payments.Add(payment);
+
+                OnCoinRemoved?.Invoke(this, EventArgs.Empty);
+
+                return payments.ToArray();
+            }
+
+            public void Pay(params ICoin[] payments)
+            {
+                payments = payments.OrderByDescending(coin => coin.Value).Take(RemainingCapacity).ToArray();
+
+                coins.AddRange(payments);
+
+                OnCoinAdded?.Invoke(this, EventArgs.Empty);
+            }
```

Comment:
Wording could be improved here to better contextualize the method actions. Charge and Pay do not feel like opposites to me. I think using (admittedly boring) words like "Withdraw" and "Deposit" or "Remove" and "Add" for method names could provide more immediate context. I see where you are going with using "payments" to mean the coins set aside for a transaction, and I can't think of something better at the moment, but my lizard brain has a hard time scanning over these methods and intuiting the meaning of that word.